src/briefcase/console.py

```diff
@@ -45,122 +92,163 @@ def boolean_input(self, question, default=False):
-                result = transform(result)
-
-            if result in choices:
-                return result
 
-            print()
-            print(error_message)
+        def is_valid_choice(user_input: str, choices: set):
+            """Validator for text_input() for a specific range of integers."""
+            if user_input in choices:
+                return True
+            raise ValueError("Enter a number from {start} to {end}".format(start=min(choices), end=max(choices)))
+
+        if intro:
+            self.print(intro)
+            self.print()
+
+        # normalize the options in to a list of 2-item tuples
+        if isinstance(options, dict):
+            ordered = list(
+                sorted(
+                    options.items(),
+                    key=operator.itemgetter(1)
+                )
+            )
+        elif all(isinstance(item, (tuple, list)) for item in options):
+            ordered = options
+        else:
+            ordered = [(item, item) for item in options]
 
-    def text_input(self, prompt, default=None):
+        # display the options to the user
+        adjusted_default = None
+        for i, (key, value) in enumerate(ordered, start=1):
+            self.print('  {i}) {label}'.format(i=i, label=value))
+            # recalculate the default for text_input
+            if default is not None and key == default:
+                adjusted_default = str(i)
+        # ensure a default is not passed to text_input unless it was actually found in the options
+        default = adjusted_default
+        self.print()
+
+        # while the internal value of the choices being considered are integers, making them
+        # strings avoids corner cases where strings could be confusingly coerced to integers
+        # for purposes of validation. For instance, if "9.0" is provided to text_input, then
+        # technically int("9.0") (which returns 9) would be considered valid for >=9 choices.
+        validator = partial(is_valid_choice, choices=set(map(str, range(1, len(ordered)+1))))
+        choice = self.text_input(prompt=prompt, input_name=input_name, default=default, validator=validator)
+        return ordered[int(choice) - 1][0]
+
+    def text_input(self, intro=None, prompt="", input_name=None, default=None, validator=None):
         """
         Prompt the user for text input.
 
-        If no default is specified, the input will be returned as entered.
+        If intro is specified, it will be presented once prior to the input prompt.
+
+        If prompt is not specified, a prompt of the form "{input_name} [{default}}: "
+            will be shown to the user.
+        If prompt is specified, it is shown to the user and input_name is ignored.
+
+        If validator is not specified:
+            the user's input is returned.
+            If the user enters nothing, the default value is returned if specified,
+                else an empty string is returned.
```

Comment:
When the documentation for your method contains multiple levels of if clauses, it's time to reconsider whether the design is correct :-)

Reply:
Reverted the input handling refactor.